src/OverlayRendering.cpp

```diff
@@ -0,0 +1,271 @@
+#include "OverlayRendering.h"
+#include "Prelude.h"
+
+extern uint8_t r_image_font_6_11[2464];
+
+/*
+ * Context for the software renderer.
+ */
+struct bitmap
+{
+    uint32_t  w, h;
+    uint32_t *pixel;
+};
+
+static void r_puts(struct bitmap *r, uint32_t sx, uint32_t sy, const char *text);
+static void r_line(struct bitmap *r, uint32_t x0, uint32_t y0, uint32_t x1, uint32_t y1, uint32_t colour);
+
+/*
+ * Draw text to bitmap.
+ * The order of drawing a character is right to left.
+ * NOTE: Cannot start drawing with negative x and y values.
+ */
+void
+r_puts(struct bitmap *r, uint32_t sx, uint32_t sy, const char *text)
+{
+	uint32_t       j      = 0;
+	const uint32_t colour = 0xffffffff;
+
+	if (sy >= r->h || text == NULL)
+	{
+		return;
+	}
+
+	while (text[j])
+	{
+		uint32_t   i;
+		const char ch = text[j];
+
+		if (ch >= 32 && ch <= 126)
+		{
+			for (i = 0; i < 11 && sy + i < r->h; i++)
+			{
+				unsigned char test = '?';
+				unsigned char mask = 1;
+				uint32_t      x    = sx + 7;
+
+				test = r_image_font_6_11[(uint32_t)(ch - 32) * 11 + i];
+
+				while (mask != 0)
+				{
+					const uint32_t map = (sy + i) * r->w + (x);
+
+					/* Must check for each iteration as x is unsigned and may underflow. */
+					if (x < r->w)
+					{
+						r->pixel[map] = (test & mask ? colour : r->pixel[map]);
+					}
+
+					x--;
+					mask <<= 1;
+				}
+			}
+		}
+		sx += 6;
+		j++;
+	}
+}
+
+/*
+ * Bresenham's line algorithm.
+ * Assuming positions are within the extent of the bitmap (screen).
+ * NOTE: Offscreen positions are handled by clamping to the extent of the bitmap,
+ * this is only meant for drawing lines within the screen.
+ */
+void
+r_line(struct bitmap *r, uint32_t x0, uint32_t y0, uint32_t x1, uint32_t y1, uint32_t colour)
+{
+	int32_t dx,  dy;
+	int32_t sx,  sy;
+	int32_t err, e2;
+
+    /*
+     * Clamp values to the extent of the bitmap.
+     * Probably extraneous as positions should be valid in the first place,
+     * else something is horribly wrong (with the caller).
+     * Clamping should be done using linear interpolation (lerp)... but this should be removed anyway...
+     * a*(1.0f-x)+(b*x), a, b : range, x : value.
+     */
+	if (x0 > r->w) { x0 = r->w; }
+	if (x1 > r->w) { x1 = r->w; }
+	if (y0 > r->h) { y0 = r->h; }
+	if (y1 > r->h) { y1 = r->h; }
+
+	dx  = abs((int32_t) x1 - (int32_t) x0);
+	sx  = (x0 > x1 ? 1 : -1);
+	dy  = abs((int32_t) y1 - (int32_t) y0);
+	sy  = (y0 > y1 ? 1 : -1);
+	err = (dx > dy ? dx : -dy) / 2;
+
+	for (;;)
+	{
+		if (x0 < r->w && y0 < r->h)
+		{
+			r->pixel[x0 + y0 * r->w] = colour;
+		}
+		if (x0 == x1 && y0 == y1)
+		{
+			break;
+		}
+		
+		e2 = err;
+		if (e2 > -dx)
+		{
+			err -= dy;
+			x0 = (uint32_t) ((int32_t) x0 - sx);
+		}
+		if (e2 < dy)
+		{
+			err += dx;
+			y0 = (uint32_t) ((int32_t) y0 - sy);
+		}
+	}
+}
+
+void
```

Comment:
please don't use a different code style than anywhere else in the engine - don't separate the return type into its own line. if the return type is too complex, make a `using` alias. If not, there's no need for it to have its own line.